src/entity_resolver/core/graph.py

```diff
@@ -1,44 +1,188 @@
+        eng_stopwords = stopwords.words("english")
+        doc = [w for w in doc if w not in eng_stopwords]
+        if to_stem:
+            ps = PorterStemmer()
+            ps_stems = []
+            for word in doc:
+                ps_stems.append(ps.stem(word))
+            return ps_stems
+        else:
+            return doc
+
+    def _clean_person_name(self, value):
+        """
+        :param value: name string e.g. 'S.D. Whitehead' or 'Whitehead, S.D.'
+        :return: list of first name and last name parts e.g. ['whitehead', 's.d.']
+        """
+        value = value.lower()
+        if ',' in value:
+            first, last = value.split(',', 1)
+            first = first.strip()
+            last = ''.join(''.join(last.split('.')).split())
+            return [first, last]
+        else:
+            names = value.split()
+            first = names[-1]
+            if len(names) == 1:
+                last = ''
+            else:
+                last = ''.join(names[:-1]).strip()
+            return [first, last]
+
+
 
 
 class Node:
 
-    def __init__(self, node_id, attrs):
+    def __init__(self, node_id, edge, attrs):
+        """
+        :param node_id: int
+        :param edge: edge object
+        :param attrs:  list of Attribute()
+        """
         self.id = node_id
-        self.attrs = attrs
-        pass
+        self.edge = edge
+        self.attrs = {}
+        for attr in attrs:
+            self.attrs[attr.name] = attr
 
     def __hash__(self):
         return self.id
 
     def __eq__(self, other):
         return self.id == other.id
 
+    def get_attr_names(self):
+        '''
+        :return: list of attribute names
+        '''
+        return list(self.attrs.keys())
 
-class Edge:
+    def get_attr(self, name):
+        '''
+        :param name: name of attribute
+        :return: list of tokenized words if atttribute type is string
+        '''
+        attr = self.attrs[name]
+        return attr.value
 
-    def __init__(self, edge_id, nodes, attrs):
+
+class Edge:
+    def __init__(self, edge_id, attrs=None):
+        """
+        :param edge_id: int
+        :param attrs: list of Attribute()
+        """
         self.id = edge_id
-        self.nodes = nodes
+        self.nodes = []
         self.attrs = attrs
-        pass
+
+    def add_node(self, node):
+        """
+        :param node: Node() object
+        """
+        self.nodes.append(node)
 
 
 class Graph:
 
-    def __init__(self, nodes, edges):
-        self.nodes = nodes
-        self.edges = edges
-        pass
+    def __init__(self, nodes, edges, attr_types):
+        """
+        :param nodes: iterables (list) of Node()
+        :param edges: dictionary of edges with key edge_id and value Edge()
+        """
+        self.nodes = list(nodes)
+        self.edges = list(edges.values())
+        self.edge_dict = edges
+        self.attr_types = attr_types
+        self.id2node = {}
+        self.id2edge = {}
```

Comment:
Both `id2node` and `id2edge` seem redundant. They can also be removed from `add_nodes` and `add_edges`